### ISSUE TITLE
sets

### ISSUE THREAD

**Issue by LinusU** [apiaryio/mson#55]
How about having sets, which is really just arrays that can have predefined values at most one time?

Proposed syntax:

```md
+ privileges (set[string], required)
    + network
    + mailman
    + procmod
    + procview
    + lpadmin
    + lpoperator
```

**Comment by zdne** [comment]
```apib
+ privileges (array[string], fixed)
    + network
    + mailman
    + procmod
    + procview
    + lpadmin
    + lpoperator
```

Is what you looking at. Do not forget that the `required` there makes `privileges` key mandatory in its parent object.

**Comment by zdne** [comment]
`fixed` is essentially saying the instance of this type should appear **exactly** as defined below.

**Comment by zdne** [comment]
oh wait, you want just one value to be there?

**Comment by zdne** [comment]
or any number of those values? but only the values from the set?

**Comment by zdne** [comment]
Either way I think it might have a solution with combination of `array` `enum` and maybe `fixed`

**Comment by LinusU** [comment]
I want the following objects to be considered valid:

```js
{ "privileges": [] }
{ "privileges": [ "procmod" ] }
{ "privileges": [ "procmod", "lpoperator", "network" ] }
```

And the following to be considered invalid:

```js
{}
{ "privileges": [ "somethingelse" ] }
{ "privileges": [ "procmod", "procmod", "network" ] }
```

**Comment by honzajavorek** [comment]
I personally understand "set" as something like array, but with unique
members and unsorted.

The uniqueness could be theoretically achieved by a validation rule. The
fact it's unsorted... not sure how much that's important in web API
contracts.

**Comment by LinusU** [comment]
@zdne Does "s1: awaiting input" mean that you are waiting on more information from me? I'm happy to help or provide more info 👍

**Comment by pksunkara** [comment]
@LinusU I think we are waiting for more feedback from other users here.

I also wonder how JSON Schema handles this.

**Comment by LinusU** [comment]
JSON Schema:

```json
{
  "type": "array",
  "uniqueItems": true,
  "items": {
    "enum": [ "network", "mailman", "procmod", "procview", "lpadmin", "lpoperator" ]
  }
}
```

**Comment by zdne** [comment]
@LinusU 

> I want the following objects to be considered valid
> ```
> { "privileges": [] }
> { "privileges": [ "procmod" ] }
> { "privileges": [ "procmod", "lpoperator", "network" ] }
> ```

@LinusU in that case, this should be the solution: 

```
# Permitted Values (enum[string))
- procmod
- lpoperator
- network

# My Object
- set (array[Permitted Values])
```

Does it work?

**Comment by LinusU** [comment]
@zdne That would work for the happy case, but then this array would also be considered valid:

```js
[ "procmod", "procmod" ]
```

**Comment by zdne** [comment]
@LinusU fair point. I think we have a case for `set` or a new type modifier (`unique`)?
